cmd/nerdctl/container/container_create_linux_test.go

```diff
@@ -174,3 +180,92 @@ func TestCreateWithTty(t *testing.T) {
 	withTtyContainer := base.InspectContainer(withTtyContainerName)
 	assert.Equal(base.T, 0, withTtyContainer.State.ExitCode)
 }
+
+// TestIssue2993 tests https://github.com/containerd/nerdctl/issues/2993
+func TestIssue2993(t *testing.T) {
+	testutil.DockerIncompatible(t)
+	t.Parallel()
+
+	getAddrHash := func(addr string) string {
+		const addrHashLen = 8
+
+		d := digest.SHA256.FromString(addr)
+		h := d.Encoded()[0:addrHashLen]
+
+		return h
+	}
+
+	type testCase struct {
+		description, containerName, namespace string
+		command                               func(t *testing.T, containerName, namespace, containersPath, etchostsPath string)
+	}
+
+	testCases := []testCase{
+		{
+			description:   "Issue #2993 - nerdctl no longer leaks containers and etchosts directories and files when container creation fails.",
+			containerName: "failure-of-container-creation",
+			namespace:     "failure-of-container-creation",
+			command: func(t *testing.T, containerName, namespace, containersPath, etchostsPath string) {
+				base := testutil.NewBaseWithNamespace(t, namespace)
+
+				defer base.Cmd("rm", "-f", containerName).Run()
```

Comment:
Use t.Cleanup instead of defer.
Also make sure to call your cleanup routine ALSO before you start doing anything to take care of leftovers from previous tests.

Reply:
Thanks !!! fixed based on [the testing principles](https://github.com/containerd/nerdctl/blob/main/docs/testing/tools.md).